django_project/api/api_views/v2/clustering.py

```diff
@@ -0,0 +1,86 @@
+# coding=utf-8
+__author__ = 'Anita Hapsari <anita@kartoza.com>'
+__date__ = '09/04/19'
+
+from django.contrib.gis.geos import Polygon, MultiPoint
+from rest_framework import serializers
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from sklearn.cluster import KMeans
+from pandas import DataFrame
+import numpy as np
+from localities_osm.utilities import get_all_osm_query
+from localities_osm.models.locality import LocalityOSMView
+
+
+class Clustering(APIView):
+
+    def get(self, request):
+        query = get_all_osm_query()
+
+        bbox = request.GET.get('bbox', None)
+        if bbox:
+            bbox_array = bbox.split(',')
+            bbox_array = [float(i) for i in bbox_array]
+            bbox_geom = Polygon.from_bbox(bbox_array)
+            query = query.filter(geometry__within=bbox_geom)
+
+        serializer = LocalitiesSerializer(query, many=True)
+        df = DataFrame(
+            serializer.data,
+            columns=['centroid_lat', 'centroid_lng'], index=None)
+
+        response_data = []
+        total = query.count()
+
+        if total > 0:
+            clusters = 20
+            if total < 20:
+                clusters = total
+
+            kmeans = KMeans(n_clusters=clusters, max_iter=50).fit(df)
+            samples = kmeans.predict(df)
+            centroids = kmeans.cluster_centers_
+
+            for i in range(len(centroids)):
+                response_data.append({
+                    'centroid_lat': centroids[i][0],
+                    'centroid_lng': centroids[i][1],
+                    'count': len([j for j in samples if j == i])
+                })
+
+            response = {
+                'data': response_data,
+                'cluster': 'true',
+                'total_data': query.count(),
+                'mean_distance': kmeans.inertia_
+            }
+
+        else:
+            response = {
+                'data': response_data,
+                'cluster': 'false',
+                'total_data': query.count()
+            }
+
+        return Response(response)
+
+
+class LocalitiesSerializer(serializers.ModelSerializer):
```

Comment:
please move this into locality_osm/serializer
renamed into
```
LocalityOSMCentroidSerializer
```

and maybe just use latitude, longitude fields